### PR TITLE
ARTEMIS-1474 TimedBuffer improved doc and refactored dead brenches on methods

### PR DESCRIPTION
**Pull request by franz1981**
It contains:
 - refactored (but semantically equals) logic 
 - improved docs on code

### CODE REVIEW

**Comment by franz1981** [comment]
@clebertsuconic And old PR of mine has changed the timeout error detection: this PR is aiming to reuse the previous logic while maintaining the adaptive (discounted) sleep.

File: artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/buffer/TimedBuffer.java [hunk 1]
```diff
@@ -34,6 +34,25 @@
...
 
 public final class TimedBuffer {
+
+   /**
+    * Property name to set the percentage of error allowed while expiring the flush {@code timeout} to happen:
+    * it can assume any positive value from {@code 0} to {@link Integer#MAX_VALUE}.
+    * <p>
+    * By default it is {@link #DEFAULT_TIMEOUT_ERROR_PERCENTAGE} more than the configured {@code timeout}.
+    */
+   public static final String JOURNAL_TIMEOUT_ERROR_PROPERTY_NAME = "journal.timeout.error";
```
Comment (clebertsuconic) [inline artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/buffer/TimedBuffer.java]: Why do we need another System property?

Comment (clebertsuconic) [reply artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/buffer/TimedBuffer.java]: I would remove the System.property here...
If this doesn't make sense in the configuration.. probably it's something users won't need to configure.

Comment (franz1981) [reply artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/buffer/TimedBuffer.java]: You're right: TBH it was more for development purposes...
Next week I've planned to do some benchmarks with persistence enabled and I was thinking that would be nice to have it customizable :P
How it looks the rest of the refactored logic?

File: artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/buffer/TimedBuffer.java [hunk 2]
```diff
@@ -34,6 +34,25 @@
...
+   public static final int DEFAULT_TIMEOUT_ERROR_PERCENTAGE = 50;
+   private static final double MAX_TIMEOUT_ERROR;
+
+   static {
+      final int errorPercentage = Integer.getInteger(JOURNAL_TIMEOUT_ERROR_PROPERTY_NAME, DEFAULT_TIMEOUT_ERROR_PERCENTAGE);
+      if (errorPercentage < 0) {
+         throw new RuntimeException("The sleep error percentage must be >= 0");
+      }
+      MAX_TIMEOUT_ERROR = 1 + (errorPercentage / 100);
```
Comment (clebertsuconic) [inline artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/buffer/TimedBuffer.java]: What is this timeout error? what magic number is this?

Before the error was a constant of 50% of the requested nanoTime... I don't understand what is this.

Comment (franz1981) [reply artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/buffer/TimedBuffer.java]: ahah is not a magic number: it is the same as before :)
The old one was `0,5` ( ie `50%`):
`max timeout error = 1,5 = 1 + 0,5 = 1 + (50/100)`

Comment (franz1981) [reply artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/buffer/TimedBuffer.java]: In the old code was indeed:
```
               // I'm letting the real time to be up to 50% than the requested sleep.
               if (realTimeSleep > timeout * 1.5) {
                  failedChecks++;
               }
```

Comment (clebertsuconic) [reply artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/buffer/TimedBuffer.java]: @franz1981 ok, what is the point of this PR? to have a parameterized parameter to tweak about? 

I don't want that kind of property open... users can do weird things and then we have to support it.. so.. keep it a constant here.... this is about making sure sleep is working or not.


on which case, the semantic will be same as before.. so I'm not understanding what is the point of the change?

**Comment by clebertsuconic** [comment]
I still don't understand the point of this PR...

but in case we go ahead, please keep it as:


```java
public final class TimedBuffer {

   private static final double MAX_TIMEOUT_ERROR_FACTOR = 1.5;
```

and use that factor instead of those properties. Keep it Simple!

**Comment by franz1981** [comment]
@clebertsuconic 
The point of the PR is to restore the original timeout error detection but use the new discounted sleep (it is the description of the PR!): I can remove the property for the reasons you mentioned, while the rest of code refactoring was done to simplify and/or make explicit/document the existing logic.

**Comment by clebertsuconic** [comment]
let me do some tests on this.. don't merge it just yet

**Pull request closed without merging**